modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java

```diff
@@ -1517,20 +1518,40 @@ private long configuredOffheap() {
      */
     private void topologySnapshotMessage(IgniteClosure<String, Void> clo, long topVer, DiscoCache discoCache,
         int evtType, ClusterNode evtNode, int srvNodesNum, int clientNodesNum, int totalCpus, double heap,
-        double offheap) {
+        double offheap, boolean needNodesDetails) {
         DiscoveryDataClusterState state = discoCache.state();
 
-        String summary = PREFIX + " [" +
-            (discoOrdered ? "ver=" + topVer + ", " : "") +
-            "locNode=" + U.id8(discoCache.localNode().id()) +
-            ", servers=" + srvNodesNum +
-            ", clients=" + clientNodesNum +
-            ", state=" + (state.active() ? "ACTIVE" : "INACTIVE") +
-            ", CPUs=" + totalCpus +
-            ", offheap=" + offheap + "GB" +
-            ", heap=" + heap + "GB]";
-
-        clo.apply(summary);
+        SB summary = new SB(PREFIX);
+
+        summary.a(" [");
+        summary.a(discoOrdered ? "ver=" + topVer + ", " : "");
+        summary.a("locNode=").a(U.id8(discoCache.localNode().id()));
+        summary.a(", servers=").a(srvNodesNum);
+        summary.a(", clients=").a(clientNodesNum);
+        summary.a(", state=").a(state.active() ? "ACTIVE" : "INACTIVE");
+        summary.a(", CPUs=").a(totalCpus);
+        summary.a(", offheap=").a(offheap).a("GB");
+        summary.a(", heap=").a(heap).a("GB");
+
+        if (evtType == EVT_NODE_JOINED && needNodesDetails) {
+            summary.a(", nodes=[");
+
+            String prefix = "";
+
+            for (ClusterNode clusterNode : discoCache.allNodes()) {
```

Comment:
I think we should print only alive nodes. You can check it by `DiscoCache#alive`. Don't forget to update prefix `", nodes=["` too.